packages/shared/scope-hoisting/src/shake.js

```diff
@@ -97,12 +108,30 @@ function isPure(path) {
   return path.isPure();
 }
 
-function isExportAssignment(path) {
+function hasSideEffects(binding) {
+  let {node} = binding.path;
+  if (isVariableDeclarator(node)) {
+    return !(
+      (!binding.referenced || isObjectExpression(node.init)) &&
+      (binding.constant ||
+        binding.constantViolations.every(
+          ({node}) =>
+            !isAssignmentExpression(node) || isObjectExpression(node.right),
+        ))
+    );
+  }
+
+  return false;
+}
+
+function isExportAssignment(path, exportsMap: Map<Symbol, Asset>) {
   let {parent} = path;
-  // match "path.foo = bar;"
+  // match "exports.foo = bar;" for known exports
```

Comment:
The reason why I wrote the comment `path.foo = bar;` is that I was never quite certain if the paremeter `path` is the `exports.foo` or the `exports` or the `bar` part or the whole assignment. Could you somehow keep that information?